File: 03-Azure/01-04-AI/01_Agentic_SDLC/src/api-cs/Controllers/OrderDetailsController.cs
using Microsoft.AspNetCore.Mvc;
using OctoSupply.Api.Models;
using OctoSupply.Api.Repositories;
using OctoSupply.Api.Utils;

namespace OctoSupply.Api.Controllers;

[ApiController]
[Route("api/order-details")]
public sealed class OrderDetailsController(OrderDetailsRepository repository) : ControllerBase
{
    private readonly OrderDetailsRepository _repository = repository;

    [HttpGet]
    public async Task<ActionResult<IReadOnlyList<OrderDetail>>> GetAll() =>
        Ok(await _repository.FindAllAsync());

    [HttpGet("{id:int}")]
    public async Task<ActionResult<OrderDetail>> GetById(int id)
    {
        var item = await _repository.FindByIdAsync(id);
        return item is null ? throw new NotFoundException("OrderDetail", id) : Ok(item);
    }

    [HttpPost]
    public async Task<ActionResult<OrderDetail>> Create([FromBody] CreateOrderDetailRequest request)
    {
        if (request.OrderId <= 0 || request.ProductId <= 0)
        {
            throw new ValidationException("orderId and productId are required");
        }

        var created = await _repository.CreateAsync(request);
        return CreatedAtAction(nameof(GetById), new { id = created.OrderDetailId }, created);
    }

    [HttpPut("{id:int}")]
    public async Task<ActionResult<OrderDetail>> Update(int id, [FromBody] UpdateOrderDetailRequest request) =>
        Ok(await _repository.UpdateAsync(id, request));

    [HttpDelete("{id:int}")]
    public async Task<IActionResult> Delete(int id)
    {
        await _repository.DeleteAsync(id);
        return NoContent();
    }
}

File: 03-Azure/01-04-AI/01_Agentic_SDLC/src/api-java/src/main/java/com/octocat/supply/controller/OrderDetailsController.java
package com.octocat.supply.controller;

import com.octocat.supply.exception.NotFoundException;
import com.octocat.supply.model.OrderDetailModel.CreateOrderDetailRequest;
import com.octocat.supply.model.OrderDetailModel.OrderDetail;
import com.octocat.supply.model.OrderDetailModel.UpdateOrderDetailRequest;
import com.octocat.supply.repository.OrderDetailsRepository;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/order-details")
public class OrderDetailsController {
    private final OrderDetailsRepository repository;

    public OrderDetailsController(OrderDetailsRepository repository) {
        this.repository = repository;
    }

    @GetMapping
    public List<OrderDetail> getAll() {
        return repository.findAll();
    }

    @GetMapping("/{id}")
    public OrderDetail getById(@PathVariable int id) {
        return repository.findById(id).orElseThrow(() -> new NotFoundException("OrderDetail not found"));
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public OrderDetail create(@RequestBody CreateOrderDetailRequest request) {
        ControllerValidation.requirePositive(request.orderId(), "orderId");
        ControllerValidation.requirePositive(request.productId(), "productId");
        return repository.create(request);
    }

    @PutMapping("/{id}")
    public OrderDetail update(@PathVariable int id, @RequestBody UpdateOrderDetailRequest request) {
        repository.findById(id).orElseThrow(() -> new NotFoundException("OrderDetail not found"));
        repository.update(id, request);
        return repository.findById(id).orElseThrow(() -> new NotFoundException("OrderDetail not found"));
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void delete(@PathVariable int id) {
        if (!repository.delete(id)) {
            throw new NotFoundException("OrderDetail not found");
        }
    }
}
